File: termin/visualization/render/components/light_component.py
from __future__ import annotations

from typing import Optional

import numpy as np

from termin.visualization.core.python_component import PythonComponent
from termin.editor.inspect_field import InspectField
from termin.lighting import LightType, Light, LightShadowParams


class LightComponent(PythonComponent):
    """
    Простейший компонент источника света.
    Хранит тип, цвет, интенсивность и параметры теней.
    """

    inspect_fields = {
        "light_type": InspectField(
            path="light_type",
            label="Light Type",
            kind="enum",
            choices=[
                (LightType.DIRECTIONAL, "Directional"),
                (LightType.POINT, "Point"),
                (LightType.SPOT, "Spot"),
            ],
            setter=lambda self, v: setattr(self, 'light_type', v if isinstance(v, LightType) else LightType(int(v))),
        ),
        "color": InspectField(
            path="color",
            label="Color",
            kind="color",
        ),
        "intensity": InspectField(
            path="intensity",
            label="Intensity",
            kind="float",
        ),
        "shadows_enabled": InspectField(
            path="shadows_enabled",
            label="Cast Shadows",
            kind="bool",
        ),
        "shadows_map_resolution": InspectField(
            path="shadows_map_resolution",
            label="Shadow Resolution",
            kind="int",
            min=256,
            max=4096,
            step=256,
        ),
        "cascade_count": InspectField(
            path="cascade_count",
            label="Cascade Count",
            kind="int",
            min=1,
            max=4,
            step=1,
        ),
        "max_distance": InspectField(
            path="max_distance",
            label="Max Distance",
            kind="float",
            min=1.0,
            max=1000.0,
            step=10.0,
        ),
        "split_lambda": InspectField(
            path="split_lambda",
            label="Split Lambda",
            kind="float",
            min=0.0,
            max=1.0,
            step=0.1,
        ),
        "cascade_blend": InspectField(
            path="cascade_blend",
            label="Cascade Blend",
            kind="bool",
        ),
    }

    def __init__(
        self,
        light_type: LightType = LightType.DIRECTIONAL,
        color=(1.0, 1.0, 1.0),
        intensity: float = 1.0,
        shadows: Optional[LightShadowParams] = None,
        enabled: bool = True,
    ):
        super().__init__()
        self.enabled = enabled
        self.light_type = light_type
        self.color = np.asarray(color, dtype=np.float32)
        self.intensity = float(intensity)
        self._shadows = shadows if shadows is not None else LightShadowParams()

    @property
    def shadows_enabled(self) -> bool:
        return self._shadows.enabled

    @shadows_enabled.setter
    def shadows_enabled(self, value: bool):
        self._shadows.enabled = value

    @property
    def shadows_map_resolution(self) -> int:
        return self._shadows.map_resolution

    @shadows_map_resolution.setter
    def shadows_map_resolution(self, value: int):
        self._shadows.map_resolution = int(value)

    @property
    def shadows(self) -> LightShadowParams:
        return self._shadows

    @property
    def cascade_count(self) -> int:
        return self._shadows.cascade_count

    @cascade_count.setter
    def cascade_count(self, value: int):
        self._shadows.cascade_count = int(value)

    @property
    def max_distance(self) -> float:
        return self._shadows.max_distance

    @max_distance.setter
    def max_distance(self, value: float):
        self._shadows.max_distance = float(value)

    @property
    def split_lambda(self) -> float:
        return self._shadows.split_lambda

    @split_lambda.setter
    def split_lambda(self, value: float):
        self._shadows.split_lambda = float(value)

    @property
    def cascade_blend(self) -> bool:
        return self._shadows.cascade_blend

    @cascade_blend.setter
    def cascade_blend(self, value: bool):
        self._shadows.cascade_blend = bool(value)

    def to_light(self) -> Light:
        """Преобразовать в объект Light для рендеринга."""
        return Light(
            type=self.light_type,
            color=self.color,
            intensity=self.intensity,
            shadows=self._shadows,
        )

File: cpp/termin/lighting/light_component.cpp
#include "light_component.hpp"
#include "../entity/component_registry.hpp"

#ifdef TERMIN_HAS_NANOBIND
#include "tc_inspect.hpp"
#else
#include "tc_inspect_cpp.hpp"
#endif

namespace termin {

LightComponent::LightComponent() {
    link_type_entry("LightComponent");
}

std::string LightComponent::get_light_type_str() const {
    return light_type_to_string(light_type);
}

void LightComponent::set_light_type_str(const std::string& type) {
    light_type = light_type_from_string(type);
}

Light LightComponent::to_light() const {
    Light l;
    l.type = light_type;
    l.color = color;
    l.intensity = intensity;
    l.shadows = shadows;

    if (entity().valid()) {
        // Direction from entity transform (forward axis = +Y)
        auto pose = entity().transform().global_pose();
        // Forward vector in Y-forward convention
        Vec3 forward = pose.ang.rotate(Vec3(0.0, 1.0, 0.0));
        l.direction = forward;
        l.position = pose.lin;
    }

    return l;
}

REGISTER_COMPONENT(LightComponent, CxxComponent);

// Register fields with choices for light_type
static struct _LightTypeFieldRegistrar {
    _LightTypeFieldRegistrar() {
        tc::InspectFieldInfo info;
        info.type_name = "LightComponent";
        info.path = "light_type";
        info.label = "Light Type";
        info.kind = "string";

        info.choices.push_back({"directional", "Directional"});
        info.choices.push_back({"point", "Point"});
        info.choices.push_back({"spot", "Spot"});

        info.getter = [](void* obj) -> tc_value {
            auto* c = static_cast<LightComponent*>(obj);
            return tc_value_string(c->get_light_type_str().c_str());
        };

        info.setter = [](void* obj, tc_value value, tc_scene_handle) {
            auto* c = static_cast<LightComponent*>(obj);
            if (value.type == TC_VALUE_STRING && value.data.s) {
                c->set_light_type_str(value.data.s);
            } else if (value.type == TC_VALUE_INT) {
                // Legacy: old scenes store light_type as int (0=Directional, 1=Point, 2=Spot)
                c->light_type = static_cast<LightType>(value.data.i);
            }
        };

        tc::InspectRegistry::instance().add_field_with_choices("LightComponent", std::move(info));
    }
} _light_type_registrar;

// Register color field
static struct _LightColorFieldRegistrar {
    _LightColorFieldRegistrar() {
        tc::InspectFieldInfo info;
        info.type_name = "LightComponent";
        info.path = "color";
        info.label = "Color";
        info.kind = "color";

        info.getter = [](void* obj) -> tc_value {
            auto* c = static_cast<LightComponent*>(obj);
            tc_vec3 v = {c->color.x, c->color.y, c->color.z};
            return tc_value_vec3(v);
        };

        info.setter = [](void* obj, tc_value value, tc_scene_handle) {
            auto* c = static_cast<LightComponent*>(obj);
            if (value.type == TC_VALUE_VEC3) {
                c->color = Vec3(value.data.v3.x, value.data.v3.y, value.data.v3.z);
            } else if (value.type == TC_VALUE_LIST && tc_value_list_size(&value) == 3) {
                // JSON stores color as [r, g, b] array
                tc_value* r = tc_value_list_get(&value, 0);
                tc_value* g = tc_value_list_get(&value, 1);
                tc_value* b = tc_value_list_get(&value, 2);
                auto get_double = [](tc_value* v) -> double {
                    if (!v) return 0.0;
                    if (v->type == TC_VALUE_DOUBLE) return v->data.d;
                    if (v->type == TC_VALUE_FLOAT) return v->data.f;
                    if (v->type == TC_VALUE_INT) return static_cast<double>(v->data.i);
                    return 0.0;
                };
                c->color = Vec3(get_double(r), get_double(g), get_double(b));
            }
        };

        tc::InspectRegistry::instance().add_field_with_choices("LightComponent", std::move(info));
    }
} _light_color_registrar;

// Register intensity field
INSPECT_FIELD(LightComponent, intensity, "Intensity", "double", 0.0, 100.0, 0.1)

// Register shadow fields
static struct _LightShadowFieldsRegistrar {
    _LightShadowFieldsRegistrar() {
        // shadows_enabled
        {
            tc::InspectFieldInfo info;
            info.type_name = "LightComponent";
            info.path = "shadows_enabled";
            info.label = "Cast Shadows";
            info.kind = "bool";

            info.getter = [](void* obj) -> tc_value {
                auto* c = static_cast<LightComponent*>(obj);
                return tc_value_bool(c->shadows.enabled);
            };

            info.setter = [](void* obj, tc_value value, tc_scene_handle) {
                auto* c = static_cast<LightComponent*>(obj);
                if (value.type == TC_VALUE_BOOL) {
                    c->shadows.enabled = value.data.b;
                } else if (value.type == TC_VALUE_INT) {
                    c->shadows.enabled = value.data.i != 0;
                }
            };

            tc::InspectRegistry::instance().add_field_with_choices("LightComponent", std::move(info));
        }

        // shadows_map_resolution
        {
            tc::InspectFieldInfo info;
            info.type_name = "LightComponent";
            info.path = "shadows_map_resolution";
            info.label = "Shadow Resolution";
            info.kind = "int";
            info.min = 256;
            info.max = 4096;
            info.step = 256;

            info.getter = [](void* obj) -> tc_value {
                auto* c = static_cast<LightComponent*>(obj);
                return tc_value_int(c->shadows.map_resolution);
            };

            info.setter = [](void* obj, tc_value value, tc_scene_handle) {
                auto* c = static_cast<LightComponent*>(obj);
                if (value.type == TC_VALUE_INT) {
                    c->shadows.map_resolution = static_cast<int>(value.data.i);
                } else if (value.type == TC_VALUE_FLOAT) {
                    c->shadows.map_resolution = static_cast<int>(value.data.f);
                } else if (value.type == TC_VALUE_DOUBLE) {
                    c->shadows.map_resolution = static_cast<int>(value.data.d);
                }
            };

            tc::InspectRegistry::instance().add_field_with_choices("LightComponent", std::move(info));
        }

        // cascade_count
        {
            tc::InspectFieldInfo info;
            info.type_name = "LightComponent";
            info.path = "cascade_count";
            info.label = "Cascade Count";
            info.kind = "int";
            info.min = 1;
            info.max = 4;
            info.step = 1;

            info.getter = [](void* obj) -> tc_value {
                auto* c = static_cast<LightComponent*>(obj);
                return tc_value_int(c->shadows.cascade_count);
            };

            info.setter = [](void* obj, tc_value value, tc_scene_handle) {
                auto* c = static_cast<LightComponent*>(obj);
                if (value.type == TC_VALUE_INT) {
                    c->shadows.cascade_count = static_cast<int>(value.data.i);
                } else if (value.type == TC_VALUE_FLOAT) {
                    c->shadows.cascade_count = static_cast<int>(value.data.f);
                } else if (value.type == TC_VALUE_DOUBLE) {
                    c->shadows.cascade_count = static_cast<int>(value.data.d);
                }
            };

            tc::InspectRegistry::instance().add_field_with_choices("LightComponent", std::move(info));
        }

        // max_distance
        {
            tc::InspectFieldInfo info;
            info.type_name = "LightComponent";
            info.path = "max_distance";
            info.label = "Max Distance";
            info.kind = "float";
            info.min = 1.0;
            info.max = 1000.0;
            info.step = 10.0;

            info.getter = [](void* obj) -> tc_value {
                auto* c = static_cast<LightComponent*>(obj);
                return tc_value_float(c->shadows.max_distance);
            };

            info.setter = [](void* obj, tc_value value, tc_scene_handle) {
                auto* c = static_cast<LightComponent*>(obj);
                if (value.type == TC_VALUE_FLOAT) {
                    c->shadows.max_distance = value.data.f;
                } else if (value.type == TC_VALUE_DOUBLE) {
                    c->shadows.max_distance = static_cast<float>(value.data.d);
                } else if (value.type == TC_VALUE_INT) {
                    c->shadows.max_distance = static_cast<float>(value.data.i);
                }
            };

            tc::InspectRegistry::instance().add_field_with_choices("LightComponent", std::move(info));
        }

        // split_lambda
        {
            tc::InspectFieldInfo info;
            info.type_name = "LightComponent";
            info.path = "split_lambda";
            info.label = "Split Lambda";
            info.kind = "float";
            info.min = 0.0;
            info.max = 1.0;
            info.step = 0.1;

            info.getter = [](void* obj) -> tc_value {
                auto* c = static_cast<LightComponent*>(obj);
                return tc_value_float(c->shadows.split_lambda);
            };

            info.setter = [](void* obj, tc_value value, tc_scene_handle) {
                auto* c = static_cast<LightComponent*>(obj);
                if (value.type == TC_VALUE_FLOAT) {
                    c->shadows.split_lambda = value.data.f;
                } else if (value.type == TC_VALUE_DOUBLE) {
                    c->shadows.split_lambda = static_cast<float>(value.data.d);
                } else if (value.type == TC_VALUE_INT) {
                    c->shadows.split_lambda = static_cast<float>(value.data.i);
                }
            };

            tc::InspectRegistry::instance().add_field_with_choices("LightComponent", std::move(info));
        }

        // cascade_blend
        {
            tc::InspectFieldInfo info;
            info.type_name = "LightComponent";
            info.path = "cascade_blend";
            info.label = "Cascade Blend";
            info.kind = "bool";

            info.getter = [](void* obj) -> tc_value {
                auto* c = static_cast<LightComponent*>(obj);
                return tc_value_bool(c->shadows.cascade_blend);
            };

            info.setter = [](void* obj, tc_value value, tc_scene_handle) {
                auto* c = static_cast<LightComponent*>(obj);
                if (value.type == TC_VALUE_BOOL) {
                    c->shadows.cascade_blend = value.data.b;
                } else if (value.type == TC_VALUE_INT) {
                    c->shadows.cascade_blend = value.data.i != 0;
                }
            };

            tc::InspectRegistry::instance().add_field_with_choices("LightComponent", std::move(info));
        }
    }
} _light_shadow_fields_registrar;

} // namespace termin
